### PR TITLE
httpc: add url_encode and url_decode methods

### LINKED ISSUE #3682: URL percent-encoding
cURL easy API contains functions `curl_easy_escape` and `curl_easy_unescape` that are respectively analogues of `url_encode` and `url_decode`:

* https://curl.haxx.se/libcurl/c/curl_easy_escape.html
* https://curl.haxx.se/libcurl/c/curl_easy_unescape.html

Old implementation of these functions in `http.server` are slow and inefficient.

RFC: https://www.notion.so/tarantool/http-add-percent-encoding-decoding-of-query-string-in-request-76a2425a4c4744a1a72643527a4fe7f7

### PR DESCRIPTION
**Pull request by olegrok**
Lua FFI bindings of curl_easy_escape and curl_easy_unescape.

Closes #3682

### CODE REVIEW

**Comment by akudiyar** [comment]
I suppose having a native Lua implementation of URL encode/decode functions and improving those of http.server. The advantage is the absence of a not-necessary cURL dependency and no ffi call which cannot be optimized by LuaJIT (and may have potential leaks).

**Comment by olegrok** [comment]
@akudiyar 

It seems you don't understand how LuaJIT and FFI works:

Please, read LuaJIT and [FFI](http://luajit.org/ext_ffi.html) documentation before saying that "ffi call which cannot be optimized by LuaJIT"

>The FFI library is tightly integrated into LuaJIT (it's not available as a separate module). The code generated by the JIT-compiler for accesses to C data structures from Lua code is on par with the code a C compiler would generate. Calls to C functions can be inlined in JIT-compiled code, unlike calls to functions bound via the classic Lua/C API.

Moreover if tarantool depends on cURL is ok to use it's this way

**Comment by akudiyar** [comment]
@olegrok

Sorry, my bad, I missed the ffi integration into LuaJIT, nvm.

**Review by Totktonada — COMMENTED** [review]
Performance numbers for url encode: https://gist.github.com/Totktonada/e29523da679cf85f52d2d7fb681d373d

Before we choose one certain implementation, please describe what specific task you want to solve with this PR. I'm not sure we want to stick with RFC 3986 or RFC 1738 or application/x-www-form-urlencoded encoding. Maybe some more general solution is needed. Maybe we should not stick with a normal form of one encoding and choose intersection on unreserved symbols: and so being more or less compatible with both standards. Maybe we need to provide ability to encode/decode with '+' as a space.

It also worth to research what methods are provided in other languages (if any).

W/o at least one certain task we have no ground under foots to make decisions.

File: src/lua/string.lua [hunk 1]
```diff
@@ -410,6 +418,53 @@ local function string_rstrip(inp, chars)
...
+-- @function url_encode
+-- @string       inp    the string
+-- @returns             result string or nil
+local function string_url_encode(inp)
+    if type(inp) ~= 'string' then
+        error(err_string_arg:format(1, "string.url_encode", 'string', type(inp)), 2)
+    end
+    local handle = ffi.C.curl_easy_init()
+    if not handle then
+        return nil
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: What error we should report to user when curl_easy_init failed?

I saw the example https://curl.haxx.se/libcurl/c/curl_easy_escape.html. They don't use curl_easy_cleanup after fail

File: src/lua/string.lua [hunk 2]
```diff
@@ -423,3 +478,5 @@ string.fromhex    = string_fromhex
 string.strip      = string_strip
 string.lstrip      = string_lstrip
 string.rstrip      = string_rstrip
+string.url_encode  = string_url_encode
+string.url_decode  = string_url_decode
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Ok. Move to http.client

File: src/lua/string.lua [hunk 3]
```diff
@@ -410,6 +418,53 @@ local function string_rstrip(inp, chars)
...
+        return nil
+    end
+
+    local escaped_str = ffi.C.curl_easy_escape(handle, inp, #inp)
+    if escaped_str == nil then
+        return nil
+    end
+
+    escaped_str = ffi.gc(escaped_str, ffi.C.curl_free)
+    return ffi.string(escaped_str)
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Done

File: src/lua/string.lua [hunk 4]
```diff
@@ -410,6 +418,53 @@ local function string_rstrip(inp, chars)
...
+local function string_url_decode(inp)
+    if type(inp) ~= 'string' then
+        error(err_string_arg:format(1, "string.url_decode", 'string', type(inp)), 2)
+    end
+    local handle = ffi.C.curl_easy_init()
+    if not handle then
+        return nil
+    end
+
+    local outlength = ffi.new("int[1]")
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Done

File: src/lua/string.lua [hunk 5]
```diff
@@ -20,6 +20,14 @@ local memcmp  = ffi.C.memcmp
 local memmem  = ffi.C.memmem
 local isspace = ffi.C.isspace
 
+ffi.cdef([[
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Done. Check please, is it correct and enough?

File: src/lua/string.lua [hunk 6]
```diff
@@ -20,6 +20,14 @@ local memcmp  = ffi.C.memcmp
...
+ typedef void CURL;
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Done

File: test/app-tap/string.test.lua [hunk 7]
```diff
@@ -370,4 +370,19 @@ test:test("unicode", function(test)
     test:is(utf8.upper(''), '', 'upper empty')
 end)
 
+test:test("url_encode/url_decode", function(test)
+    test:plan(8)
+    test:is('hello', string.url_encode('hello'), 'correct encoding of eng')
+    test:is('%D0%BF%D1%80%D0%B8%D0%B2%D0%B5%D1%82', string.url_encode('привет'), 'correct encoding of rus')
```
[the remark this answers is not in the document]
Comment (olegrok) [reply test/app-tap/string.test.lua]: Done. Check, is it enough?

**Comment by olegrok** [comment]
@Totktonada 

> Performance numbers for url encode: https://gist.github.com/Totktonada/e29523da679cf85f52d2d7fb681d373d
> 
> Before we choose one certain implementation, please describe what specific task you want to solve with this PR. I don't sure we want to stick with RFC 3986 or RFC 1738 or application/x-www-form-urlencoded encoding. Maybe some more general solution is needed. Maybe we should not stick with a normal form of one encoding and choose intersection on unreserved symbols: and so being more or less compatible with both standards. Maybe we need to provide ability to encode/decode with '+' as a space.
> 
> It also worth to research what methods are provided in other languages (if any).
> 
> W/o at least one certain task we have no ground under foots to make decisions.

Python - https://docs.python.org/3/library/urllib.parse.html As I see they try to change old versions of RFC to 3986

Javascript - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeuricomponent : RFC 3986

Ruby - https://bugs.ruby-lang.org/issues/2542 : Change to RFC 3986

curl - https://curl.haxx.se/libcurl/c/curl_easy_escape.html (On the bottom on page in "See also" RFC 3986)

I don't sure that we need to support old specifications. Let's support only RFC 3986. If user wants to support legacy systems, don't touch existing custom code. This feature doesn't break anything

However, there is deprecated functions: https://curl.haxx.se/libcurl/c/curl_escape.html and https://curl.haxx.se/libcurl/c/curl_upescape.html that implement RFC 2396

File: src/lua/string.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: Add curl_easy_cleanup

File: src/lua/string.lua [hunk 8]
```diff
@@ -410,6 +418,53 @@ local function string_rstrip(inp, chars)
...
+        error(err_string_arg:format(1, "string.url_encode", 'string', type(inp)), 2)
+    end
+    local handle = ffi.C.curl_easy_init()
+    if not handle then
+        return nil
+    end
+
+    local escaped_str = ffi.C.curl_easy_escape(handle, inp, #inp)
+    if escaped_str == nil then
+        return nil
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/string.lua]: I don't know, how make the best, I return error message as second argument.

`'curl_easy_init error', 'curl_easy_escape error', 'curl_easy_unescape error'`

**Comment by Totktonada** [comment]
> Javascript - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeuricomponent : RFC 3986

Please, read the link you paste. It is not RFC 3986 compliant. (I didn't check other links.)

> I don't sure that we need to support old specifications.

application/x-www-form-urlencoded is the actual standard (based on RFC 1738). It just other then 'url encoding'.

Again, what is the specific problem you try to solve here?

**Comment by olegrok** [comment]
> Again, what is the specific problem you try to solve here?

Not copy-paste code from one project to another when I want to make http request with query parameters. In particular to our http.server. I think if we have http.client on-board to make it more complete we should have some utils functions too

**Comment by olegrok** [comment]
> Please, read the link you paste. It is not RFC 3986 compliant. (I didn't check other links.)

Oh, yes, you are right. 

In one place they say that don't escape "A-Z a-z 0-9 - _ . ! ~ * ' ( )"

Bottom they say:
> To be more stringent in adhering to RFC 3986 ...

**Comment by Totktonada** [comment]
> > Again, what is the specific problem you try to solve here?
>
> Not copy-paste code from one project to another when I want to make http request with query parameters. In particular to our http.server. I think if we have http.client on-board to make it more complete we should have some utils functions too

As I see http.server needs to at least decode application/x-www-form-urlencoded data and the functions you propose don't cover this case. It worth to think whether we can provide a general solution that will handle all related cases.

There is also application/x-www-form-urlencoded-like encoding in one of our closed project, but it doesn't correspond neither to RFC 3986, not application/x-www-form-urlencoded (this is the restriction of an external API). I want to hope that such cases are rare, but JS example shows that it can be common. That is why we should look deeply into other implementations and cases when an implementation provides several functions (like PHP) or configurable ones.

**Comment by olegrok** [comment]
> > > Again, what is the specific problem you try to solve here?
> > 
> > 
> > Not copy-paste code from one project to another when I want to make http request with query parameters. In particular to our http.server. I think if we have http.client on-board to make it more complete we should have some utils functions too
> 
> As I see http.server needs to at least decode application/x-www-form-urlencoded data and the functions you propose don't cover this case. It worth to think whether we can provide a general solution that will handle all related cases.
> 
> There is also application/x-www-form-urlencoded-like encoding in one of our closed project, but it doesn't correspond neither to RFC 3986, not application/x-www-form-urlencoded (this is the restriction of an external API). I want to hope that such cases are rare, but JS example shows that it can be common. That is why we should look deeply into other implementations and cases when an implementation provides several functions (like PHP) or configurable ones.

One thing that I found:
https://curl.haxx.se/libcurl/c/curl_url_get.html and https://curl.haxx.se/libcurl/c/curl_url_set.html that could be used for urls parsing. They say that follow "RFC 3986+" (as I understand they change space to "+")

Curl can't provide us with something except "RFC 3986" and I didn't find similar functions in another libraries which tarantool depends. Extension support is beyond the scope of such a task "Add curl bindings for urlescape/urlunescape"

**Comment by Totktonada** [comment]
> Not copy-paste code from one project to another when I want to make http request with query parameters. In particular to our http.server. I think if we have http.client on-board to make it more complete we should have some utils functions too

I misread it before and was think that you want these function to use in the http.server implementation. Now I understand that you want to add query parameters to an http request with http.client. I think it is better to support passing query parameters as a table and encapsulate all that questions about encodings inside.

**Comment by Totktonada** [comment]
BTW, found interesting article on the topic: https://www.talisman.org/~erlkonig/misc/lunatech%5Ewhat-every-webdev-must-know-about-url-encoding/

**Comment by kostja** [comment]
No news since May 21. Please work on the requested changes and resubmit the pull request.

**Pull request closed without merging**